service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,244 @@
+func isInVersionSet(tbid string, sets []*persistencepb.CompatibleVersionSet) bool {
+	for _, set := range sets {
+		for _, bid := range set.BuildIds {
+			if bid.GetId() == tbid {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func InsertAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_InsertBuildIdAssignmentRule,
+	maxARs int) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		data = &persistencepb.VersioningData{AssignmentRules: make([]*persistencepb.AssignmentRule, 0)}
+	} else {
+		data = common.CloneProto(data)
+	}
+	if req.GetRuleIndex() < 0 {
+		return nil, serviceerror.NewInvalidArgument("rule index cannot be negative")
+	}
+	persistenceAR := persistencepb.AssignmentRule{
+		Rule:            req.GetRule(),
+		CreateTimestamp: timestamp,
+		DeleteTimestamp: nil,
+	}
+	slices.Insert(data.GetAssignmentRules(), int(req.GetRuleIndex()), &persistenceAR)
+	return data, checkAssignmentConditions(data, maxARs, false)
+}
+
+func ReplaceAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceBuildIdAssignmentRule) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		return nil, serviceerror.NewInvalidArgument("data cannot be nil for a replace call")
+	} else {
+		data = common.CloneProto(data)
+	}
+	rules := data.GetAssignmentRules()
+	idx := int(req.GetRuleIndex())
+	if len(rules) > idx+1 {
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("rule index %d is out of bounds for assignment rule list of length %d", idx, len(rules)))
+	}
+	persistenceAR := persistencepb.AssignmentRule{
+		Rule:            req.GetRule(),
+		CreateTimestamp: timestamp,
+		DeleteTimestamp: nil,
+	}
+	slices.Replace(data.AssignmentRules, idx, idx+1, &persistenceAR)
+	return data, checkAssignmentConditions(data, 0, req.GetForce())
+}
+
+func DeleteAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteBuildIdAssignmentRule) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		return nil, serviceerror.NewInvalidArgument("data cannot be nil for a delete call")
+	} else {
+		data = common.CloneProto(data)
+	}
+	rules := data.GetAssignmentRules()
+	idx := int(req.GetRuleIndex())
+	if len(rules) > idx+1 {
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("rule index %d is out of bounds for assignment rule list of length %d", idx, len(rules)))
+	}
+	rule := rules[idx]
+	rule.DeleteTimestamp = timestamp
+	return data, checkAssignmentConditions(data, 0, req.GetForce())
+}
+
+// checkRedirectConditions returns an error if there is no room for the new Redirect Rule, or if it causes the
+// queue to fail the other requirements
+func checkRedirectConditions(g *persistencepb.VersioningData, maxRRs int) error {
+	rules := g.GetRedirectRules()
+	if maxRRs > 0 && len(rules) > maxRRs {
```

Comment:
this check also needs to exclude tombstoned rules. but let's not worry about redirect rules in this PR.